From daca416415180b6a6a6d0b894627435dde81d0ae Mon Sep 17 00:00:00 2001
From: Jaehoon Lee <jaehoon85.lee@lge.com>
Date: Thu, 28 Jan 2021 15:02:02 +0900
Subject: [PATCH] Fix v4l2dec flush for video loop playback

:Release Notes:
Fix video loop playback

:Detailed Notes:
Fix video loop playback issue
caused by flush error in v4l2 decoder
when get EOS.

:Testing Performed:
Tested in raspberrypi4

:QA Notes:

:Issues Addressed:
[PLAT-127537] [OSE] Video loop attribute
              does not work for ose
---
 .../sys/v4l2/gstv4l2videodec.c                | 22 ++++++-------------
 1 file changed, 7 insertions(+), 15 deletions(-)

diff --git a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2videodec.c b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2videodec.c
index b3d4b693c1..19cd025a70 100644
--- a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2videodec.c
+++ b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2videodec.c
@@ -331,37 +331,29 @@ gst_v4l2_video_dec_flush (GstVideoDecoder * decoder)
 
   GST_DEBUG_OBJECT (self, "Flushed");
 
-  self->output_flow = GST_FLOW_FLUSHING;
-
   /* Ensure the processing thread has stopped for the reverse playback
    * discount case */
   if (gst_pad_get_task_state (decoder->srcpad) == GST_TASK_STARTED) {
+    GST_VIDEO_DECODER_STREAM_UNLOCK (decoder);
     gst_v4l2_object_unlock (self->v4l2output);
     gst_v4l2_object_unlock (self->v4l2capture);
     gst_pad_stop_task (decoder->srcpad);
+	GST_VIDEO_DECODER_STREAM_LOCK (decoder);
   }
 
+  self->output_flow = GST_FLOW_OK;
+
   if (G_UNLIKELY (!g_atomic_int_get (&self->active)))
     return TRUE;
 
   gst_v4l2_object_unlock_stop (self->v4l2output);
   gst_v4l2_object_unlock_stop (self->v4l2capture);
 
-  /* Ensure our internal pools are flushed properly*/
   if (self->v4l2output->pool)
-    gst_buffer_pool_set_active (self->v4l2output->pool, FALSE);
-
-  if (self->v4l2capture->pool)
-    gst_buffer_pool_set_active (self->v4l2capture->pool, FALSE);
-
-  self->output_flow = GST_FLOW_OK;
-
-  /* Ensure our internal pools are reactivated properly*/
-  if (self->v4l2output->pool)
-    gst_buffer_pool_set_active (self->v4l2output->pool, TRUE);
-
+    gst_v4l2_buffer_pool_flush (self->v4l2output->pool);
+  
   if (self->v4l2capture->pool)
-    gst_buffer_pool_set_active (self->v4l2capture->pool, TRUE);
+    gst_v4l2_buffer_pool_flush (self->v4l2capture->pool);
 
   return TRUE;
 }
